cirq/sim/act_on_args.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Objects and methods for acting efficiently on a state tensor."""
+
+from typing import Any, Iterable, Dict, Sequence
+
+import numpy as np
+
+from cirq import protocols, ops
+from cirq.protocols.decompose_protocol import _try_decompose_into_operations_and_qubits
+
+
+class ActOnArgs:
```

Comment:
Why do we prefer inheritance here over composition?

Reply:
We get to eliminate the duplicate bit shift and logging code in MeasurementGate because they are all instances of ActOnArgs

Reply:
Doing this in #4979, #5065